src/main/java/spoon/support/compiler/jdt/JDTBasedSpoonCompiler.java

```diff
@@ -401,55 +380,90 @@ protected JDTBatchCompiler createBatchCompiler(InputType... types) {
 	}
 
 	protected boolean buildTemplates(JDTBuilder jdtBuilder) {
-		if (templates.getAllJavaFiles().isEmpty()) {
-			return true;
-		}
+		CompilationUnitDeclaration[] units = buildUnits(jdtBuilder, templates, getTemplateClasspath(), "template ", false);
 
-		JDTBatchCompiler batchCompiler = createBatchCompiler(InputType.FILES);
+		// here we build the model in the template factory
+		buildModel(units);
+
+		return probs.size() == 0;
+	}
+
+	private static final CompilationUnitDeclaration[] EMPTY_RESULT = new CompilationUnitDeclaration[0];
+
+	protected CompilationUnitDeclaration[] buildUnits(JDTBuilder jdtBuilder, SpoonFolder sourcesFolder, String[] classpath, String debugMessagePrefix, boolean buildOnlyOutdatedFiles) {
+		List<SpoonFile> sourceFiles = sourcesFolder.getAllJavaFiles();
+		if (sourceFiles.isEmpty()) {
+			return EMPTY_RESULT;
+		}
+		FileCompiler batchCompiler = (FileCompiler) createBatchCompiler(InputType.FILES);
+		//tell the compiler which files it should build
+		batchCompiler.setInpuFiles(sourceFiles);
 

```

Comment:
Here is the main bug fix, but I think my changes fixes more bugs. I am sorry that I did so many changes at once, but I needed these changes to simplify code to understand it. I believe I can simplify it even more, but I preffer to have your feedback in the middle of way, to not produce something what is not acceptable for you.